Footnote support

lindeer opened the pull request:
implementation referred to [markdown-it-footnote](https://github.com/markdown-it/markdown-it-footnote/blob/master/index.js) and [pandoc](https://pandoc.org/MANUAL.html#footnotes)

lindeer commented:
hi @srawlins , could you please have a review for footnote support?

chenzhiguang commented:
I created a footnote reference specification, it has some slightly different from some of the implementations. But at least there are clear rules: 🙂

https://github.com/chenzhiguang/dart_markdown/wiki/footnote-reference-specification

chenzhiguang commented:
Test cases:
https://github.com/chenzhiguang/dart_markdown/blob/master/test/extensions/footnote_reference.unit

lindeer commented:
thanks @chenzhiguang , I think footnote is an import markdown feature, especially in article writing though not in spec, not understand why official library do not support it, nearly every other language md library has it!  But I find your implementation has huge difference from markdown-it and pandoc, and your based dart-lang md library has also great changes. I have a little concern for future maintenance.

srawlins commented:
I think the primary spec we would follow (barring commonmark, since that does not support footnotes) would be GitHub (via https://github.blog/changelog/2021-09-30-footnotes-now-supported-in-markdown-fields/). I do really like pandoc, and think it is a popular project. I have not heard of markdown-it. So if we land an implementation it will have to be compatible with GitHub with a good array of tests.

The good array of tests is needed since GitHub does not have a spec for their footnote support (😢 ). That is still a bummer because we just have to... manually test with Babelmark or something? 😢

lindeer commented:
Fine @srawlins , I would refine the code according to github's [example](https://docs.github.com/en/get-started/writing-on-github/getting-started-with-writing-and-formatting-on-github/basic-writing-and-formatting-syntax#footnotes). The main difference is paragraph making in footnote-definition part, and I will add more test cases.

srawlins commented on lib/src/inline_syntaxes/footnote_ref_syntax.dart:
```diff
@@ -0,0 +1,32 @@
+import 'inline_syntax.dart';
+
+class FootnoteRefSyntax extends InlineSyntax {
+  FootnoteRefSyntax() : super(r'!?\[\^(\S+?)\]');
```
Hmm, what is the leading `!?` for?

lindeer replied:
refine the code and not extends but directly change within `LinkSyntax`

srawlins commented on test/extensions/footnote_block.unit:
```diff
@@ -0,0 +1,80 @@
+>>> footnote with paragraph
```
Also it looks like leading whitespace is allowed in the ref?

```markdown
Footnote 3 link[^ a].

[^a]: Text
```

works in GitHub.

lindeer replied:
now supported

srawlins commented on test/extensions/footnote_block.unit:
```diff
@@ -0,0 +1,80 @@
+>>> footnote with paragraph
```
Can we add a test where the footnote def is first?

```markdown
[^a]: Text

Footnote 3 link[^a].
```

srawlins commented on lib/src/inline_syntaxes/footnote_ref_syntax.dart:
```diff
@@ -0,0 +1,32 @@
+  bool onMatch(InlineParser parser, Match match) {
+    final label = match[1]!;
+    final refs = parser.document.footnoteReferences;
+    if (!refs.containsKey(label)) {
```
Since `footnoteReferences` is typed with non-nullable types, we can simplify some of this code by grabbing the value first with `final something = parser.document.footnoteReferences[label];`. Then we can guard this if-statement with `if (something != null)` and line 19 can be `parser.document.footnoteReferences[label] = something + 1;`

srawlins commented on lib/src/inline_syntaxes/footnote_ref_syntax.dart:
```diff
@@ -0,0 +1,32 @@
+    refs[label] = (refs[label] ?? 0) + 1;
+    final pos = refs.keys.toList(growable: false).indexOf(label);
+
+    if (parser.charAt(parser.start) == $exclamation) {
```
Why is an exclamation handled separately here?

lindeer replied:
code  refined, and not need any more.

srawlins commented on lib/src/document.dart:
```diff
@@ -65,6 +66,8 @@ class Document {
   List<Node> parseLines(List<String> lines) {
     final nodes = BlockParser(lines, this).parseLines();
     _parseInlineContent(nodes);
+    // after parsing inline as we need ref count
```
Make sure to start comments with a capiital, and end with a period. https://dart.dev/guides/language/effective-dart/documentation#do-format-comments-like-sentences

lindeer replied:
done

srawlins commented on lib/src/document.dart:
```diff
@@ -84,6 +87,41 @@ class Document {
+      ..clear()
+      ..addAll(blocks);
+
+    final section = footnotes.isEmpty
```
Can these few lines be simplified by guarding them all with `footnotes.isEmpty` rather than doing a second guard below with `section != null`.

lindeer replied:
change guarding to `isEmpty`

srawlins commented on lib/src/document.dart:
```diff
@@ -84,6 +87,41 @@ class Document {
+            _classElement('ol', 'footnotes-list', footnotes),
+          ]);
+    if (section != null) {
+      nodes.add(_classElement('hr', 'footnotes-divider', null));
```
I don't think we need the `null`.

lindeer replied:
done also

srawlins commented on lib/src/document.dart:
```diff
@@ -84,6 +87,41 @@ class Document {
     }
   }
+
+  static Element _classElement(
```
This name is a bit odd; maybe something like `_elementWithClass`.

Alternatively, this helper may not be needed with my suggestions below; `Element(tag, children)..attributes['class'] = foo` should be fine to read.

srawlins reviewed (changes requested):
Thanks for the PR!

srawlins commented:
@lindeer thanks for the new commit! Since it was force-pushed and overwrote the previous content, I can't see what was changed. Could you go through my comments and say something like "Done" on each if it has been addressed, and click "Resolve conversation" on those. I think I had a question or two as well which you could answer inline (I don't think I was specifically asking for code comments...) Thanks!

lindeer commented on test/extensions/footnote_block.unit:
```diff
@@ -0,0 +1,80 @@
+<p>and multiple paragraphs. <a href="#fnref1" class="footnote-backref">↩︎</a></p>
+</li>
+</ol>
+</section>
```
done

lindeer commented:
sorry srawlins, I squashed commits too early, and github not addressed automatically. I think most important part of new commits was the introducing of 'reparse' in InlineParser. When failed to match link, we check if current content match footnote between '[' and ']', "[...][“ and "]", this result in that some leading content lost, so rollback 'pos' and 'start' to starting position and do some clearing things.  Other parts reuse link syntax logic, trying to do minimum changes.

And I added many corner cases, some different from github (I commented `unlike github`), but that's because of our own design.

lindeer commented:
hi @srawlins ,I found my way of 'reparse' implementation was  not elegant, we could modify `close()` of  `LinkSyntax`  without any change of `InlineParser`. So could I drop this PR and create 2 separated PRs?
1. first one for modifying existing  code without any regression.
2. after the first was  merged, the second rebased onto the first,  and was to implement footnote.

This way could make code modification more clear to review, could I?

srawlins commented:
@lindeer That's not a bad idea. But you don't have to close this one first. If you create a PR for modifying `LinkSyntax.close`, I can see how it would benefit this PR.

srawlins commented:
Hi @lindeer I see you pushed last week. Is this ready for review again? I've lost context, apologies.

lindeer commented:
@srawlins yes, it is. I just rebased onto new main branch, and added a new rebase fix, but github listed all commits.

srawlins commented on lib/src/document.dart:
```diff
@@ -14,6 +14,10 @@ import 'util.dart';
 /// Maintains the context needed to parse a Markdown document.
 class Document {
   final Map<String, LinkReference> linkReferences = {};
+  // keys are case sensitive.
```
Nit: Start with a capital letter [1] and hyphenate 'case-sensitive'.

[1] https://dart.dev/guides/language/effective-dart/documentation#do-format-comments-like-sentences

srawlins commented on lib/src/document.dart:
```diff
@@ -14,6 +14,10 @@ import 'util.dart';
   final Map<String, LinkReference> linkReferences = {};
+  // keys are case sensitive.
+  final footnoteReferences = <String, int>{};
+  // Keep footnotes appearing order, and labels are case insensitive.
```
Nit: Hyphenate 'case-insensitive'.

srawlins commented on lib/src/block_syntaxes/footnote_def_syntax.dart:
```diff
@@ -0,0 +1,74 @@
+import 'paragraph_syntax.dart';
+import 'setext_header_syntax.dart';
+
+class FootnoteDefSyntax extends BlockSyntax {
```
Would you mind adding a doc comment explaining the source of this syntax? Thanks!

srawlins replied:
This needs to be addressed, please.

srawlins replied:
I'm sorry, @lindeer , we cannot accept this PR unless this comment is addressed. We need to have documented where the syntax is specified. Who controls its definition / where can we go for updates, etc?

lindeer replied:
sorry @srawlins , I have no idea why I missed these comment..

lindeer replied:
I came to know that [footnotes are not in the spec yet](https://github.com/github/cmark-gfm/issues/283#issuecomment-1378868725).  It is said that there is a private repo for the spec somewhere in GH.
I found a GFM footnote grammar and its rust implementation [here](https://github.com/wooorm/markdown-rs/blob/2498e31eecead798efc649502bbf5f86feaa94be/src/construct/gfm_footnote_definition.rs).
or what other I could do to move this PR forward?

srawlins replied:
Sorry, I was on vacation for a few days 😭 but we're really almost there. I'd love to land this with like one more change 😁 . It is totally fine that there is not a published spec; that's all we have so that's ok. I do still love this feature because it helps us match what GitHub parses. So thanks again for your work!

With a feature that is under-specified like this, I think it' just important to add all of the links we can 😁 If you could add, in the doc comment links to:

* the issue: https://github.com/github/cmark-gfm/issues/283#issuecomment-1378868725
* the cmark-gfm source: https://github.com/github/cmark-gfm/blob/master/src/blocks.c#L1212
* the rust implementation is a great resource too, I think you can add it for good measure: https://github.com/wooorm/markdown-rs/blob/2498e31eecead798efc649502bbf5f86feaa94be/src/construct/gfm_footnote_definition.rs

srawlins commented on lib/src/block_syntaxes/footnote_def_syntax.dart:
```diff
@@ -0,0 +1,74 @@
+      ...parseChildLines(parser),
+    ];
+    // Linebreak in secondary paragraph did not match LineBreakSyntax
+    // so there is no `<br />`, unlike github.
```
nit: 'GitHub' should be capitalized.

srawlins commented on lib/src/block_syntaxes/footnote_def_syntax.dart:
```diff
@@ -0,0 +1,74 @@
+
+  static bool _isBlock(BlockParser parser, String line) {
+    final syntaxList = parser.blockSyntaxes
+        .takeWhile((s) => !_invalidSecondaryBlock.contains(s.runtimeType));
```
I don't want to use runtimeType here, as that would not account for (theoretical, to be sure) subtypes.

I think we can use `.whereType` and a late instance field to compute this once per parse. Something like:

```dart
late final secondaryBlockSyntaxes =
  [...parser.blockSyntaxes]
    .removeWhere((e) = _invalidSecondaryBlock.contains(e));
```

WDYT?

srawlins replied:
> seems have to use runtimeType, we actually expect a iterable with same base type `BlockSyntax` but different subtype, `whereType` could only give us a iterable with same type.

Sorry I don't understand this. You might be misunderstanding what `whereType` does. `[1, 2.1, 'hello'].whereType<num>()` will yield both `1` and `2.1`.

> And for a late instance, I use `??=` to assign a local variable, a class field would make `FootnoteDefSyntax` none const constructor any more.

OK a local variable is ok here.

lindeer replied:
I use BlockSytax's `pattern` field to exclude `EmptyBlockSyntax`, `ParagraphSyntax` and `SetextHeaderSyntax`, no `runtimeType` any more.

srawlins commented on lib/src/block_syntaxes/footnote_def_syntax.dart:
```diff
@@ -0,0 +1,74 @@
+    return children.map(Line.new).toList(growable: false);
+  }
+
+  static const _invalidSecondaryBlock = <Type>[
```
Let's make this a Set since it is only used for `.contains`.

srawlins commented on lib/src/block_syntaxes/footnote_def_syntax.dart:
```diff
@@ -0,0 +1,74 @@
+    final children = BlockParser(lines, parser.document).parseLines();
+    return Element('li', children)
+      ..attributes['id'] = 'fn-$id'
+      ..attributes['_label_'] = label;
```
Instead of a fake HTML attribute which is removed later, we can add a field to Element called `footnoteLabel` or something similar. Like how Element has a `generatedId` for header IDs.

lindeer replied:
a new field make `Element` inflating, and it is a field hardly could be used by other subtype. What about using 'data attribute' from [HTML syntax](https://developer.mozilla.org/en-US/docs/Learn/HTML/Howto/Use_data_attributes), attributes to store some extra information, just like `data-footnote-label`?

srawlins replied:
I don't think we have any memory inflation concerns for this package. It doesn't outweigh the importance of the readability of the code here. A field brings in type support, typo prevention, other forms of static analysis, possible runtime optimizations, etc.

srawlins commented on lib/src/document.dart:
```diff
@@ -97,6 +103,79 @@ class Document {
+      ..addAll(blocks);
+
+    if (footnotes.isNotEmpty) {
+      // sort footnotes by appeared order
```
Nit: start with a capital letter and end with a period.

srawlins commented on lib/src/document.dart:
```diff
@@ -97,6 +103,79 @@ class Document {
+      });
+      final list = Element('ol', footnotes);
+
+      // Ignore github's attribute: <data-footnotes>.
```
Spelling, 'GitHub'.

srawlins commented on lib/src/document.dart:
```diff
@@ -97,6 +103,79 @@ class Document {
+            Element('sup', [Text(num)])..attributes['class'] = 'footnote-ref',
+        ])
+
+          // Ignore github's attributes: <data-footnote-backref aria-label="Back to content">.
```
Spelling: 'GitHub'

srawlins commented on lib/src/inline_syntaxes/footnote_ref_syntax.dart:
```diff
@@ -0,0 +1,69 @@
+import 'link_syntax.dart' show LinkContext;
+
+// Footnotes syntax share the same heading '[' with Reference Link, this makes
+// a lot of mass, if extends `LinkSyntax`, it will have to override
```
This comment is fine for the PR, but it doesn't have to be checked in. It can be removed.

srawlins commented on lib/src/inline_syntaxes/footnote_ref_syntax.dart:
```diff
@@ -0,0 +1,69 @@
+// `_tryCreateReferenceLink`, and also need other context info, so just a
+// separated class is enough.
+class FootnoteRefSyntax {
+  static String? _footnoteLabel(InlineParser parser, String key) {
```
It doesn't look like this needs `parser`?

srawlins commented on lib/src/inline_syntaxes/footnote_ref_syntax.dart:
```diff
@@ -0,0 +1,69 @@
+  static Iterable<Node>? tryCreateFootnoteLink(
+    LinkContext context,
+    String text,
+    bool secondary,
```
Avoid positional bool parameters. https://dart.dev/guides/language/effective-dart/design#avoid-positional-boolean-parameters

srawlins commented on lib/src/inline_syntaxes/footnote_ref_syntax.dart:
```diff
@@ -0,0 +1,69 @@
+    // `label` is what footnoteReferences stored, it is case sensitive.
+    final label =
+        refs.keys.firstWhere((k) => k.toLowerCase() == key, orElse: () => '');
+    // `count != null` means footnote was valid
```
End comments with period, here and below.

srawlins commented on lib/src/patterns.dart:
```diff
@@ -50,6 +50,9 @@ final listPattern =
 final tablePattern = RegExp(
     r'^[ ]{0,3}\|?([ \t]*:?\-+:?[ \t]*\|)+([ \t]|[ \t]*:?\-+:?[ \t]*)?$');
 
+/// Just keep the same with github.
```
This comment does not describe what this pattern is. We don't have to reference GitHub.

srawlins reviewed (changes requested):
Great work! I have a few comments and then I will review further. Thanks again!

lindeer commented:
hi @srawlins , seems every time I pushed a new commit, the review came into a pause, hope this could be merged earlier :cry:

srawlins commented:
Yes, apologies. We need to figure out a different system for open source contributions.

srawlins commented on lib/src/inline_syntaxes/footnote_ref_syntax.dart:
```diff
@@ -0,0 +1,65 @@
+      return null;
+    }
+    final result = <Node>[];
+    // There are 4 cases here: ![^...], [^...], ![...][^...], [...][^...]
```
Very helpful comment

srawlins commented on lib/src/inline_syntaxes/footnote_ref_syntax.dart:
```diff
@@ -0,0 +1,65 @@
+import '../charcode.dart';
+import 'link_syntax.dart' show LinkContext;
+
+class FootnoteRefSyntax {
```
Could you add comments here about where this syntax is specified? Could just be live experiments from GitHub, if it is not spelled out in GitHub-flavored Syntax.

lindeer replied:
It is a handling branch of link syntax, but I move the code into a separated class.

srawlins replied:
Please include a more full comment here. Something like "A syntax for defining a footnote reference, as specified at..." and include a click-able link.

lindeer replied:
resolved in a8a81fc

srawlins replied:
Please address the above comment:

> Please include a more full comment here. Something like "A syntax for defining a footnote reference, as specified at..." and include a click-able link.

A click-able link would probably have a URL starting with 'https://'

lindeer replied:
footnote reference here is a handling branch of `LinkSyntax`, not a kind of syntax. I just moved these code into a separated class, should I rename the class to avoid misunderstanding?

srawlins replied:
Hmm no I don't think it needs to be renamed. Maybe it can just reference `[FootnoteDefSyntax]` so that there is a documentation link between the two. Then it's fine. Thanks!

srawlins reviewed (changes requested):
This is looking so great! Thanks so much for working on this, I love it.

lindeer commented on lib/src/block_syntaxes/footnote_def_syntax.dart:
```diff
@@ -0,0 +1,80 @@
+      ...parseChildLines(parser),
+    ];
+    // Linebreak in secondary paragraph did not match LineBreakSyntax
+    // so there is no `<br />`, unlike GitHub.
```
sorry, it's a mistake, I ran `cmark-gfm` with `--hardbreaks`, remove the comment.

lindeer commented:
@srawlins could you give another review? :pray:

srawlins commented on lib/src/document.dart:
```diff
@@ -97,6 +106,81 @@ class Document {
+
+  /// Generate backref nodes, append them to footnote definition's last child.
+  void _appendBackref(List<Node> children, String ref, int count) {
+    final refs = () sync* {
```
Unfortunately `sync*` is slow. https://github.com/dart-lang/sdk/issues/32102

Why do we need a self-executing block here? (Apologies if I suggested it 😅 )

It looks like this can be:

```dart
var refs = [
  for (var i = 0; i < count; i++)
    ...[
      Text(' '),
      Element('a', [
          Text('\u21a9'),
          if (i > 0)
            Element('sup', [Text(num)])..attributes['class'] = 'footnote-ref',
        ])
            // Ignore GitHub's attributes: <data-footnote-backref aria-label="Back to content">.
              ..attributes['href'] = '#fnref-$ref$suffix'
              ..attributes['class'] = 'footnote-backref',
    ],
];
```

OK so that is _really_ not readable, haha, so I would extract out some logic into a either a real or a fake constructor on Element, like:

```dart
extension on Element {
  static Element footnoteAnchor(String ref, int index) {
    final number = '${i + 1}';
    final suffix = i > 0 ? '-$number' : '';
    return Element('a', [
          Text('\u21a9'),
          if (index > 0)
            Element('sup', [Text(number)])..attributes['class'] = 'footnote-ref',
        ])

          // Ignore GitHub's attributes: `<data-footnote-backref aria-label="Back to content">`.
          ..attributes['href'] = '#fnref-$ref$suffix'
          ..attributes['class'] = 'footnote-backref';
  }
}
```

Then the above list mess can become:


```dart
var refs = [
  for (var i = 0; i < count; i++)
    ...[Text(' '), Element.footnoteAnchor(ref, i)],
];
```

srawlins reviewed (changes requested):
Almost there, thanks!

srawlins commented on lib/src/block_syntaxes/footnote_def_syntax.dart:
```diff
@@ -0,0 +1,78 @@
+    return children.map(Line.new).toList(growable: false);
+  }
+
+  // Patterns that would be used to decide if one line is a block
```
Please use triple-slash for doc comments (see [Effective Dart](https://dart.dev/guides/language/effective-dart/documentation#do-use--doc-comments-to-document-members-and-types)) and end with a period. (see [Effective Dart](https://dart.dev/guides/language/effective-dart/documentation#do-format-comments-like-sentences)).

srawlins commented on lib/src/block_syntaxes/footnote_def_syntax.dart:
```diff
@@ -0,0 +1,78 @@
+    dummyPattern,
+  };
+
+  // Whether this line is one kind of block, if true footnotes block should end.
```
Please use `///` here too.

srawlins commented on lib/src/block_syntaxes/footnote_def_syntax.dart:
```diff
@@ -0,0 +1,78 @@
+    late final syntaxList = parser.blockSyntaxes
+        .where((s) => !_excludingPattern.contains(s.pattern));
+
+    // Every line is footnote's children util two blank lines or a block
```
Nit: end with a period: https://dart.dev/guides/language/effective-dart/documentation#do-format-comments-like-sentences

srawlins reviewed (approved):
Thanks so much for all of the doc links. That really helps ❤️ ❤️ ❤️ 

I love it, thanks!

kevmoo commented:
We need to update the changelog and the version here, friends!

CC @lindeer @srawlins

srawlins commented:
Will do :)
